### PR TITLE
Adding user journey map p2

### PR DESCRIPTION
**Pull request by Cdub-9**
Part 2 of Adding in user journey tracking into Perfume

Added in the User Journey map structure to keep track of data

Added in basic marking functions and tests

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -19,3 +19,5 @@ export const pushTask = (cb: any): void => {
     cb();
   }
 };
+
+export const getJourneyMarkName = (mark: string): string => `user_journey_mark.${mark}`
```
Comment (Zizzamia) [inline src/utils.ts:23]: I think we are missing a unit test for this function.

File: src/userJourneyMap.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        } else { 
+            userJourneyMap.activeSteps[step] = true; 
+        }
+    })
+  }
+
+  /**
+   * removes one step from active steps
+   * 
+   * @param step
```
Comment (Zizzamia) [inline src/userJourneyMap.ts:98]: We kind of not need all this `@parm` stuff in Perfume, a good description is enough.

File: src/userJourneyMap.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+   */
+  export const removeActiveStep = (step: string) => {
```
Comment (Zizzamia) [inline src/userJourneyMap.ts:100]: This function miss a test

File: src/userJourneyMap.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+          const finalStep = steps[steps.length - 1];
+          if (userJourneyMap.finalSteps[finalStep]){
+            userJourneyMap.finalSteps[finalStep].push(key)
+          } else {
+            userJourneyMap.finalSteps[finalStep] = [key];
+          }
+        }
+      },
+    );
+    console.log(JSON.stringify(userJourneyMap))
```
Comment (Zizzamia) [inline src/userJourneyMap.ts:74]: You forgot this around.

File: src/userJourneyMap.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  /**
+   * this method allows to add new steps by passing the start mark
+   *
+   * @param startMark
+   */
+  export const addActiveSteps = (startMark: string) => {
+    const newSteps = userJourneyMap.startMarkToStepsMap[startMark] ?? [];
+    // adding the new steps to the active map
+    Object.keys(newSteps).forEach(step => {
+        if(userJourneyMap.activeSteps[step]){
```
Comment (Zizzamia) [inline src/userJourneyMap.ts:87]: What's going on with your identation, some of it is 2 steps, in other places is 4.
For Perfume, should always be 2.

File: src/userJourneyMap.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        // insert when top level end mark is not present
+        userJourneyMap.finalMarkToStepsMap[endMark] = {[startMark]: [step]};
+      } else {
+        // insert when end mark and start mark are both present
+        const currentSteps = userJourneyMap.finalMarkToStepsMap[endMark][startMark];
+        currentSteps.push(step);
+        userJourneyMap.finalMarkToStepsMap[endMark][startMark] = currentSteps;
+    }
+    });
+    console.log(JSON.stringify(userJourneyMap))
```
Comment (Zizzamia) [inline src/userJourneyMap.ts:51]: You forgot this as well

File: src/userJourneyMap.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    if (!config.userJourneySteps) {
+      return;
+    }
+
+    resetUserJourneyMap();
+
+    Object.entries<IStepConfig<string>>(config.userJourneySteps).forEach(([step, { marks }]) => {
+      const startMark = marks[0];
+      const endMark = marks[1];
+      console.log(`the start mark is: ${  startMark}`)
```
Comment (Zizzamia) [inline src/userJourneyMap.ts:35]: Eheheh and this. 🤔

File: src/userJourneyMap.ts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+      if (!userJourneyMap.finalMarkToStepsMap[endMark]) {
+        // insert when top level end mark is not present
+        userJourneyMap.finalMarkToStepsMap[endMark] = {[startMark]: [step]};
+      } else {
+        // insert when end mark and start mark are both present
+        const currentSteps = userJourneyMap.finalMarkToStepsMap[endMark][startMark];
+        currentSteps.push(step);
+        userJourneyMap.finalMarkToStepsMap[endMark][startMark] = currentSteps;
+    }
+    });
```
Comment (Zizzamia) [inline src/userJourneyMap.ts:50]: Yup this indentation is all over the place. Please give it a nice clean.

File: __tests__/userJourney/onMarkJourney.spec.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ */
+import { WP } from '../../src/constants';
+import mock from '.././_mock';
+import Perfume from '../../src/perfume'
+import { markJourney, markJourneyOnce } from '../../src/userJourney';
+import { config } from '../../src/config';
+
+import { testConfig } from '../userJourneyTestConstants';
+
+describe('markJourneyOnce', () => {
```
Comment (Zizzamia) [inline __tests__/userJourney/onMarkJourney.spec.ts:12]: you are in the `onMarkJourney` file, but here you wrote `markJourneyOnce`

File: __tests__/userJourney/markJourney.spec.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  afterEach(() => {
+    if (spy) {
+      spy.mockReset();
+      spy.mockRestore();
+    }
+  })
+
+  describe('markJourney', () => {
+    it('using the markJourneyOnce function should call WP.mark with the journey name', () => {
+        spy = jest.spyOn(WP, 'mark');
```
Comment (Zizzamia) [inline __tests__/userJourney/markJourney.spec.ts:28]: Indentations are wrong.

File: __tests__/userJourney/userJourneyMap.spec.ts [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+describe('userJourneyMap', () => {
+  let spy: jest.SpyInstance;
+
+  beforeEach(() => {
+    (WP as any) = mock.performance();
+    const perfume = new Perfume(testConfig);
+  });
+
+  describe('correctly sets each map', () => {
```
Comment (Zizzamia) [inline __tests__/userJourney/userJourneyMap.spec.ts:20]: Random empty indentations.

File: src/perfume.ts [hunk 12]
```diff
@@ -71,6 +75,12 @@ export default class Perfume {
     if (WN && WN.storage && typeof WN.storage.estimate === 'function') {
       WN.storage.estimate().then(reportStorageEstimate);
     }
+    console.log('creating new perfume')
```
Comment (Zizzamia) [inline src/perfume.ts:78]: Console.log

File: src/perfume.ts [hunk 13]
```diff
@@ -71,6 +75,12 @@ export default class Perfume {
...
+    // initializing User Journeys if present
+    if(config.userJourneys && config.userJourneySteps){
```
Comment (Zizzamia) [inline src/perfume.ts:80]: Where are the tests for this?

File: src/userJourney.ts [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+const measureJourneySteps = (endMark: string) => {
+    if (userJourneyMap.finalMarkToStepsMap[endMark]) {
+      // this is an end mark so we delete the entry
+      // Implementation coming in P3 PR
+    } else {
+      addActiveSteps(endMark);
+    }
+    config.onMarkJourney?.(endMark, Object.keys(userJourneyMap.activeSteps));
+  
+  }
```
Comment (Zizzamia) [inline src/userJourney.ts:17]: Identations are all wrong.

**Comment by Zizzamia** [comment]
@Cdub-9 I did some files cleanup after I merge this PR.
Btw, when you have a minute can you fix the `npm run lint` script, so we can have easily lint our changes going forward. I noticed is not working correctly anymore.

**Comment by Cdub-9** [comment]
@Zizzamia yes, I will fix script in next PR

**Pull request merged**